Implement execution of schema changes through a state machine.

vivekmenezes opened the pull request:
A schema change is now run through many stages, where each stage runs
as a transaction on the database. Some steps involve the uses of
LeaseManager.Publish() that publishes a new version to the schema
only after there are outstanding leases to only one version of
the schema.

Since the synchronous execution of a schema change can
fail at any time (it runs multiple transactions sequentially), a second asynchronous path is added to
move the state machine along.

closes #2845 

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3411)

<!-- Reviewable:end -->

vivekmenezes commented:
let me give this one last review

vivekmenezes commented:
Ready for review. Thanks!
I'm going to be heading out on vacation but should get online by Wednesday to start incorporating feedback on this. Thanks for this review.

tamird commented:
merge conflict and CI failed. can you ping when both are resolved?

vivekmenezes committed:
sql: Schema changes correctly rolled out in multiple steps

A schema change is run through many stages, where each stage runs
as a transaction on the database. Some steps involve the use of
LeaseManager.Publish(), that publishes a new version to the schema
while maintaining the < 2 versions invariant.

Since the synchronous execution of a schema change can
fail at any time (it runs multiple transactions sequentially that
can individually fail), a second asynchronous execution path is
added to move the state machine along in case of failures.

vivekmenezes commented:
This is ready for review.  It's still not passing all tests but I believe it's because of the general flakiness. I'd rather review this because it's going to take time.

tschottdorf commented:
try a rebase, I just merged my PR disabling some flaky tests.

petermattis commented:
Review status: 0 of 18 files reviewed at latest revision, 10 unresolved discussions.

---

*[server/server.go, line 254 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RrSvbP49rThbiOtkd:-K7RrSvbP49rThbiOtke:962617778) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/server/server.go#L254)):*
Why is this commented out? If it isn't being used, let's just delete it.

---

*[sql/create.go, line 90 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Rrju3jtN4c6rG0fSw:-K7Rrju3jtN4c6rG0fSx:-1787073205) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/create.go#L90)):*
I would have thought `addIndexMutation` should bump `NextMutationID`, and perhaps return the mutation ID it added so that it could be used below in `p.notifySchemaChange`.

---

*[sql/executor.go, line 263 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RtGKCfP8bcYT68m5w:-K7RtGKCfP8bcYT68m5x:1900032600) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/executor.go#L263)):*
Is this used for testing?

---

*[sql/plan.go, line 39 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RtWtR772EO9W8LUQB:-K7RtWtR772EO9W8LUQC:-2046320682) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/plan.go#L39)):*
`s/some of the//g`

---

*[sql/schema_changer.go, line 38 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RtlSk5o7oW5p7kiux:-K7RtlSk5o7oW5p7kiuy:1690644240) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer.go#L38)):*
Does this type need to be exported?

---

*[sql/schema_changer.go, line 53 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RtoG0j4xVcnPuXZFD:-K7RtoG1NQk6k9EwOxTB:-1442164879) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer.go#L53)):*
Spurious blank line.

---

*[sql/schema_changer.go, line 56 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Rtu4LyVATCdoMdorV:-K7Rtu4LyVATCdoMdorW:-1785611628) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer.go#L56)):*
Add `TODO(vivek)`.

---

*[sql/schema_changer.go, line 59 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Ru-Et3vuWsgzwOFlE:-K7Ru-Et3vuWsgzwOFlF:1359056388) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer.go#L59)):*
No need for the separate `timestamp` variable: `p.setTxn(txn, time.Now())`.

---

*[sql/schema_changer.go, line 314 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Rur-dqzpsNnH1iET1:-K7Rur-dqzpsNnH1iET2:1140958936) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer.go#L314)):*
Why not? Well, I see why not, but you should call out why in the comment. 

---

*[sql/schema_changer_test.go, line 113 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RvNma8Il-aa9J96QU:-K7RvNma8Il-aa9J96QV:-444740778) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer_test.go#L113)):*
What's broken?

Seems like there are a lot more code paths than what this test covers.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 18 files reviewed at latest revision, 10 unresolved discussions.

---

*[server/server.go, line 254 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RrSvbP49rThbiOtkd:-K7RwhRKiKgsjwdd3ItD:-258144061) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/server/server.go#L254)):*
uncommented. I was just trying to figure out whether my change for some reason was breaking something in core. As you know the tests have been very flaky.

---

*[sql/create.go, line 90 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Rrju3jtN4c6rG0fSw:-K7Rx9BsNXYtmIl4OaUx:774853483) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/create.go#L90)):*
We don't do it that way because you can have multiple mutations using the same mutation id. We add all the mutations and then increment it at the end.

---

*[sql/executor.go, line 263 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RtGKCfP8bcYT68m5w:-K7RxOhBGpKGcsuy8cEL:-663197540) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/executor.go#L263)):*
It's the normal code path, and is disabled in testing.

---

*[sql/plan.go, line 39 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RtWtR772EO9W8LUQB:-K7RxfaS1Gopws4xAAXp:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/plan.go#L39)):*
done

---

*[sql/schema_changer.go, line 38 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RtlSk5o7oW5p7kiux:-K7RySo6hHbhaWJEbevD:1512466402) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer.go#L38)):*
only for testing to be visible in sql_test

---

*[sql/schema_changer.go, line 53 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RtoG0j4xVcnPuXZFD:-K7RyxKh48xrygG3esaq:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer.go#L53)):*
done

---

*[sql/schema_changer.go, line 56 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Rtu4LyVATCdoMdorV:-K7RzDkp69RfbmwDH-ZV:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer.go#L56)):*
done

---

*[sql/schema_changer.go, line 59 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Ru-Et3vuWsgzwOFlE:-K7RzLXwHKI4AUdUMEDK:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer.go#L59)):*
done

---

*[sql/schema_changer.go, line 314 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Rur-dqzpsNnH1iET1:-K7S-5839ENXK6RbpkwU:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer.go#L314)):*
Done

---

*[sql/schema_changer_test.go, line 113 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7RvNma8Il-aa9J96QU:-K7S-heOcx4bneDAFDXP:-1712654970) ([raw file](https://github.com/cockroachdb/cockroach/blob/f60afce0ef51b78862e40f3fa2ecb9f909c4370f/sql/schema_changer_test.go#L113)):*
Nothing's broken, except a lot of tests in storage not due to this change. Yes, there are a lot of code paths that need testing. Let me work on that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Test coverage is a lot better now. Do let me know if you'd like to see more. I think I'd like to see more higher level testing, like running a schema change on a large table while simultaneously running operations, but I'm leaving that for a time when we implement chunking of the schema change scan, when we have a this all flowing through a ton of transactions and there is more scope for weird race conditions.

petermattis commented:
Yes, the test coverage is better now.

LGTM

---

Review status: 0 of 19 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/schema_changer.go, line 431 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WmV7VmjC_KSNiipQE:-K7WmV7WgruF_l438EBS:-714615221) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/schema_changer.go#L431)):*
Should probably mention that most schema changes are executed synchronously by the node that created the schema change and that this async mechanism is meant to catch cases where the node dies.

---

*[sql/schema_changer.go, line 466 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WnEZ1DG4Ew0c2WaYU:-K7WnEZ1DG4Ew0c2WaYV:-1243992028) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/schema_changer.go#L466)):*
Using a ticker here doesn't seem quite right. Once you've found a `schemaChanger` you should be able to determine how long to wait before attempting to process it. That is, given `schemaChangers` I can loop over the changers and determine the minimum time to wait until one of the creation times is older than `asyncSchemaChangeExecDelay`. 

---

*[sql/schema_changer.go, line 474 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WmwtoUVJb5dZQGRu9:-K7WmwtoUVJb5dZQGRuA:1727910879) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/schema_changer.go#L474)):*
Do we expect this flag to change during the lifetime of the `SchemaChangeManage`. I would have thought you could check it at the beginning of this method and never run the worker.

---

*[sql/schema_changer.go, line 481 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WmeQE6FQ7O-VHUz08:-K7WmeQE6FQ7O-VHUz09:1896341450) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/schema_changer.go#L481)):*
There are enough fields here to warrant using multiple lines for readability.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-:-K7Wm84hCBkzwjmy5PkS:-2035326571)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 14 of 19 files at r3, 1 of 4 files at r5, 2 of 4 files at r7.
Review status: 17 of 19 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[sql/alter_table.go, line 194 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx3j3siA1YQodb0q5:-K7Wx3j3siA1YQodb0q6:311652059) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/alter_table.go#L194)):*
nit: this is brittle in that the `-1` is here to reverse the `++` above; can you add a local variable to hold onto the "original" `NextMutationID` instead?

---

*[sql/backfill.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx18hDmCmZVKHwtNP:-K7Wx18hDmCmZVKHwtNQ:-1489882866) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/backfill.go#L78)):*
shouldn't this be `continue`?

---

*[sql/backfill.go, line 176 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WwoNqP6YK9g6C2zVZ:-K7WwoNqP6YK9g6C2zV_:652797026) ([raw file](https://github.com/cockroachdb/cockroach/blob/27fa5819ee62082ca0f81f9d5e9f801c8a4ceb02/sql/backfill.go#L176)):*
it doesn't seem appropriate to remove this TODO

---

*[sql/create.go, line 98 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wwk2bFL3xW7sdIPV-:-K7Wwk2bFL3xW7sdIPV0:311652059) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/create.go#L98)):*
nit: this is brittle in that the `-1` is here to reverse the `++` above; can you add a local variable to hold onto the "original" `NextMutationID` instead?

---

*[sql/drop.go, line 157 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WwSvdqUMa-HJ90W5y:-K7WwSvdqUMa-HJ90W5z:311652059) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/drop.go#L157)):*
nit: this is brittle in that the `-1` is here to reverse the `++` above; can you add a local variable to hold onto the "original" `NextMutationID` instead?

---

*[sql/executor.go, line 267 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WvvZfosWH0EBbhHz_:-K7WvvZfosWH0EBbhHza:178591327) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/executor.go#L267)):*
hoist this out of the loop or even out of the function

---

*[sql/plan.go, line 39 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WvI8uM3lbaf44la62:-K7WvI8uM3lbaf44la63:72184082) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/plan.go#L39)):*
so a SchemaChanger is now a schema change?

---

*[sql/plan.go, line 41 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WvNsJ4tLKF13YYsi4:-K7WvNsJ4tLKF13YYsi5:-1129570934) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/plan.go#L41)):*
the lease is for the table, not the schema change, right?

---

*[sql/structured.go, line 48 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WuqknxGMUTdtIuw6A:-K7WuqknxGMUTdtIuw6B:-1035645820) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/structured.go#L48)):*
`const`? also, `const`s are untyped, so this would be `const invalidMutationID = 0`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 16 of 19 files reviewed at latest revision, 13 unresolved discussions.

---

*[sql/schema_changer.go, line 431 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WmV7VmjC_KSNiipQE:-K7WvLE_98sI4P1EjAQs:1387462000) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/schema_changer.go#L431)):*
done. Added a comment to SchemaChangeManager

---

*[sql/schema_changer.go, line 466 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WnEZ1DG4Ew0c2WaYU:-K7WvYszlKsMegVokvrc:-2043506613) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/schema_changer.go#L466)):*
True, but why add all that complexity. It think this is simpler and doesn't hog any CPU.

---

*[sql/schema_changer.go, line 474 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WmwtoUVJb5dZQGRu9:-K7Wvkb3SKP-uYhf8Wkp:-1838909398) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/schema_changer.go#L474)):*
Changed to never run the worker

---

*[sql/schema_changer.go, line 481 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WmeQE6FQ7O-VHUz08:-K7WwYtgXu3SokvbffPy:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/schema_changer.go#L481)):*
done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 16 of 19 files reviewed at latest revision, 13 unresolved discussions.

---

*[sql/schema_changer.go, line 466 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WnEZ1DG4Ew0c2WaYU:-K7X2NOvgW--CITvuqJE:1877053350) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/schema_changer.go#L466)):*
"All that complexity" seems a little bit excessive. I think you'd have one function which took a `[]SchemaChanger` and returned the duration to wait.  Note that it doesn't look like you stopped the ticker which leaks a goroutine: https://golang.org/pkg/time/#NewTicker.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 16 of 19 files reviewed at latest revision, 11 unresolved discussions.

---

*[sql/schema_changer.go, line 397 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7X7YWDFB6Rb43zfBOx:-K7X7YWDFB6Rb43zfBOy:-1818059052) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c3d91fbd15d89351fe71900b7008b866613748/sql/schema_changer.go#L397)):*
I'd reword the third sentence as: `If the node dies while processing the schema change this manager acts as a backup execution mechanism.`

---

*[sql/structured.go, line 48 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WuqknxGMUTdtIuw6A:-K7X7qLj9JzBrVXgzvIs:-1819711013) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/structured.go#L48)):*
`const`s can be typed, but they frequently aren't. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 9 of 19 files reviewed at latest revision, 11 unresolved discussions.

---

*[sql/alter_table.go, line 194 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx3j3siA1YQodb0q5:-K7WzBnXb1TeFA0ZS6AM:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/alter_table.go#L194)):*
done

---

*[sql/backfill.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx18hDmCmZVKHwtNP:-K7X-NbriYsTO0WVMlqn:1612526566) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/backfill.go#L78)):*
No, but you've got a point that it's a bit weird. I've added a mutationID to the parameters passed in and use that here. I've added a comment above the for loop.

---

*[sql/backfill.go, line 176 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WwoNqP6YK9g6C2zVZ:-K7X-g4SwJ6l1mPEcPu2:1264235726) ([raw file](https://github.com/cockroachdb/cockroach/blob/27fa5819ee62082ca0f81f9d5e9f801c8a4ceb02/sql/backfill.go#L176)):*
There is an existing TODO to manage large tables above.

---

*[sql/create.go, line 98 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wwk2bFL3xW7sdIPV-:-K7X-mGEY_AJAvS_D2pD:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/create.go#L98)):*
done

---

*[sql/drop.go, line 157 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WwSvdqUMa-HJ90W5y:-K7X-njswtKQQwocLBic:677588217) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/drop.go#L157)):*
doen

---

*[sql/executor.go, line 267 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WvvZfosWH0EBbhHz_:-K7X-yCb-n1plelkwUcL:362354311) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/executor.go#L267)):*
hoisted out of the loop

---

*[sql/plan.go, line 39 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WvI8uM3lbaf44la62:-K7X0WT-YZBYtbmpLVaa:665617535) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/plan.go#L39)):*
Fixed the comment

---

*[sql/plan.go, line 41 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WvNsJ4tLKF13YYsi4:-K7X1N7GGXuTX-ETxeuH:1021854879) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/plan.go#L41)):*
fixed the comment

---

*[sql/schema_changer.go, line 397 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7X7YWDFB6Rb43zfBOx:-K7XIjU1r9nGEiTqCucI:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c3d91fbd15d89351fe71900b7008b866613748/sql/schema_changer.go#L397)):*
done

---

*[sql/structured.go, line 48 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WuqknxGMUTdtIuw6A:-K7X1RCKmZ4wmhIJ3H77:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/structured.go#L48)):*
done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r8, 7 of 8 files at r9.
Review status: 17 of 19 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[sql/backfill.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx18hDmCmZVKHwtNP:-K7XyrODbW4GsV6G6H_c:623285043) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/backfill.go#L78)):*
Hm, I don't see a comment that explains this.

---

*[sql/executor.go, line 267 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WvvZfosWH0EBbhHz_:-K7Xz21hO98m2F6lap28:-1920020101) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/executor.go#L267)):*
I don't see it?

---

*[sql/plan.go, line 39 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Xz93qP8dO2KcE_BcE:-K7Xz93qP8dO2KcE_BcF:-1286386288) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5afa3bad91dfcf7a756fc2754b58630d3b6b355/sql/plan.go#L39)):*
s/changer/changers/

---

*[sql/plan.go, line 41 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7XzIMGVheRoT6rHqYz:-K7XzIMGVheRoT6rHqZ-:-559736737) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5afa3bad91dfcf7a756fc2754b58630d3b6b355/sql/plan.go#L41)):*
changers or changes? if changers, what does "execute" mean, exactly?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 16 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/backfill.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx18hDmCmZVKHwtNP:-K7bKxpjkzQBJDye95on:-198787260) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/backfill.go#L78)):*
See above

// Collect the elements that are part of the mutation.

---

*[sql/executor.go, line 267 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WvvZfosWH0EBbhHz_:-K7bLEpsSlcXphRaEl9X:1631469539) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/executor.go#L267)):*
I hoisted it outside of the loop over the schemaChangers. I could move it further up, but I don't see a point really. There is value to be able to see it locally here.

---

*[sql/plan.go, line 39 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Xz93qP8dO2KcE_BcE:-K7bMlFVuEoTtZqL2z6S:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5afa3bad91dfcf7a756fc2754b58630d3b6b355/sql/plan.go#L39)):*
done

---

*[sql/plan.go, line 41 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7XzIMGVheRoT6rHqYz:-K7bMjHQMYxlNmIYUIAW:614633638) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5afa3bad91dfcf7a756fc2754b58630d3b6b355/sql/plan.go#L41)):*
updated the comment

---

*[sql/schema_changer.go, line 466 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7WnEZ1DG4Ew0c2WaYU:-K7i_7ZhXDC1YW1CfgyI:-700703233) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/schema_changer.go#L466)):*
Added a timer that fires when the first schema change needs processing.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 4 files at r7, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[sql/executor.go, line 264 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kMFqop-BEWCVUlEtX:-K7kMFqp61tIJhHrTuRe:873065699) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/executor.go#L264)):*
is there a condition you can add here that tells us whether there are actually any schema changes pending that you'd need to wait for? `len(planMaker.schemaChangers) > 0` perhaps?

---

*[sql/executor.go, line 285 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kU4mSVVzMxsRNWE-W:-K7kU4mTyqNAfRoPtG8U:242197379) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/executor.go#L285)):*
what can be the original result we're overwriting here? Do we really want to overwrite it?

```sql
BEGIN TRANSACTION;
CREATE TABLE newtable ...;
SELECT * FROM existingtable;
COMMIT;
```

What would happen here? I would guess that it would only process the schema change  after the commit (since only then the `txn` will be nil), which is not the order specified.

But I probably just don't understand how these things work because

```sql
BEGIN TRANSACTION;
CREATE TABLE newtable ...;
INSERT INTO newtable VALUES ...;
COMMIT;
```

wouldn't even see the new table if that were how it were set up.
Feel free to just point me somewhere to read so I can catch up.

---

*[sql/lease.go, line 54 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kVOFMkttMfwI6jF6Q:-K7kVOFNv92Zzv07I4xX:978621475) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/lease.go#L54)):*
that's a weird error string and also the name is weird - is this if a version can't be updated, or it was attempted to update it illegally? Maybe `errCannotUpdateVersion` or some variation is better, depending on the semantics (which I can't figure out from this line anyway).

---

*[sql/schema_changer.go, line 201 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kW_n2eNILeacSZs4Z:-K7kW_n2eNILeacSZs4_:-1056420040) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L201)):*
reflow the comment.

---

*[sql/schema_changer.go, line 203 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kWeIK5yHy42wdjyvT:-K7kWeIK5yHy42wdjyvU:-2147076700) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L203)):*
if this returns an error, should log it.

---

*[sql/schema_changer.go, line 219 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kWlAP6F53P79v1aAX:-K7kWlAP6F53P79v1aAY:-877380979) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L219)):*
`Apply`

---

*[sql/schema_changer.go, line 223 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kWtaVZ0U0xfOKiJEI:-K7kWtaVZ0U0xfOKiJEJ:-1973552271) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L223)):*
this looks like you should wrap the error to give it context, for example
`return util.Errorf("error purging mutation after error %s: %s", err, errPurge)`

---

*[sql/schema_changer.go, line 297 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kXdj6FG8DTfMBbD3e:-K7kXdj6FG8DTfMBbD3f:373872018) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L297)):*
the comment feels redundant (and doesn't conform to the usual style).

---

*[sql/schema_changer.go, line 299 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kZ7NziMqRknIpoh1-:-K7kZ7NziMqRknIpoh10:-1294991063) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L299)):*
can you pull those out and add a comment explaining the choices (i.e. why it's ok to be so aggressive)?

---

*[sql/schema_changer.go, line 330 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kZUaERgJzf0zSiEqr:-K7kZUaERgJzf0zSiEqs:1988332781) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L330)):*
what happens if you drop a unique column and you have to reverse? You may already have deleted some of the unique data, so you can't really undo this.
I guess that's what your comment below is about.

---

*[sql/schema_changer.go, line 376 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kZqoQDVdbCYl_ywh8:-K7kZqoQDVdbCYl_ywh9:-949073686) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L376)):*
Would be interested in seeing some tests where the transaction aborts in random places to see whether everything done is idempotent (or takes retries into account).

---

*[sql/schema_changer.go, line 437 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7k_2_mKXLVjGyl46im:-K7k_2_mKXLVjGyl46in:-1057082499) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L437)):*
can't all of these live on SchemaChangeManager to avoid the awkward changes of globals?

---

*[sql/schema_changer.go, line 478 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7k_Zt6LWIOvQL1bqQq:-K7k_Zt6LWIOvQL1bqQr:-1655322366) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L478)):*
the timer doesn't attempt to run the schema changer.

---

*[sql/schema_changer.go, line 587 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7k_T1P49M96nJbhkcZ:-K7k_T1Qz5bmuKjANcSp:-1273003546) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L587)):*
rename `creationTime -> lastUpdated`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 8 of 19 files at r3, 1 of 4 files at r7, 1 of 2 files at r8, 6 of 8 files at r9.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 15 of 19 files reviewed at latest revision, 16 unresolved discussions.

---

*[sql/executor.go, line 264 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kMFqop-BEWCVUlEtX:-K7l-HlqKi4_EvgWbUO0:1872734134) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/executor.go#L264)):*
done. Also added a comment that the boolean is used to disable execution in some tests.

---

*[sql/executor.go, line 285 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kU4mSVVzMxsRNWE-W:-K7l-cspmU4mehazwJns:-1580364124) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/executor.go#L285)):*
A valuable comment to read to better understand how schema changes get executed is the one for the version field in sql/structured.proto

BTW CREATE TABLE is not a schema change, modifiying an existing table schema using a command like CREATE INDEX is.

We definitely want to override the result under most circumstances, in particular we need it when we run

statement error duplicate key value \(c\)=\(1\) violates unique constraint "bar"
ALTER TABLE t ADD CONSTRAINT bar UNIQUE (c)

When we run the above command, the transaction with ALTER TABLE succeeds, but when the schema change gets rolled out in separate transactions it hits the violation of unique constraint error, which modifies the result. It wouldn't surprise me if under certain circumstances overriding the original result doesn't make sense, but then the user is probably running a schema change with other commands in the same transaction which is really what they shouldn't be doing. We accommodate those kinds of transactions, but I think we want to strongly discourage them.

---

*[sql/lease.go, line 54 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kVOFMkttMfwI6jF6Q:-K7lKwa907biq7FLqzXx:-1550703357) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/lease.go#L54)):*
It's used by update functions called by Publish() that would like to notify Publish() that the descriptor was not modified and it should not update the version. I've renamed this error to errDidntUpdateDescriptor

---

*[sql/schema_changer.go, line 201 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kW_n2eNILeacSZs4Z:-K7lMFRjJ_sVdvUymTZZ:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L201)):*
done

---

*[sql/schema_changer.go, line 203 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kWeIK5yHy42wdjyvT:-K7lMzdhq7bpcWW3bqnR:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L203)):*
done

---

*[sql/schema_changer.go, line 219 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kWlAP6F53P79v1aAX:-K7lN2VR2u6Yc9IF-Z1Y:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L219)):*
done

---

*[sql/schema_changer.go, line 223 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kWtaVZ0U0xfOKiJEI:-K7lN89LlDPpH1GTrBXT:374784518) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L223)):*
good point. Done

---

*[sql/schema_changer.go, line 297 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kXdj6FG8DTfMBbD3e:-K7lOtTbo_xDgZd3VUAs:614633638) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L297)):*
updated the comment

---

*[sql/schema_changer.go, line 299 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kZ7NziMqRknIpoh1-:-K7lOuy-lRgoAgKtUz8p:863177888) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L299)):*
reworded and mentioned that this is aggressive because there might be a user waiting for the schema change to complete.

---

*[sql/schema_changer.go, line 330 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kZUaERgJzf0zSiEqr:-K7lVk9mN4hv0yDB7sFQ:685714133) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L330)):*
We currently do not hit an error when dropping a unique column, but might in the future. The TODO comment below is about figuring out a better solution to what we have now.

---

*[sql/schema_changer.go, line 376 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kZqoQDVdbCYl_ywh8:-K7lcdDxaY-k_HyDSzlg:1978016269) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L376)):*
I've added some tests for IsDone, although no tests if any of these transactions fail and retry. I do intent to add in a larger acceptance test for the schema change mechanism when I make the change that splits the backfill into multiple chunks of work. #3274 

---

*[sql/schema_changer.go, line 437 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7k_2_mKXLVjGyl46im:-K7lQozC5NJJtJacvj7f:-657291241) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L437)):*
I suppose they could but then we would need to expose the Manager through some API on the server. I don't think that's worth it.

---

*[sql/schema_changer.go, line 478 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7k_Zt6LWIOvQL1bqQq:-K7ldEEbJwtdu9KgOxtS:1786868355) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L478)):*
I've moved newTimer into its own method on the manager and reworded the comment.

---

*[sql/schema_changer.go, line 587 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7k_T1P49M96nJbhkcZ:-K7lVCXOTtkedBffr5Qq:1049208777) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/schema_changer.go#L587)):*
renamed to execAfter to keep track of the time deadline after which the schema changer can execute.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
LGTM with some more comments.

---

Reviewed 4 of 4 files at r11.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 285 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kU4mSVVzMxsRNWE-W:-K7lgohg7Esvtl7nwWIK:-1767836945) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/executor.go#L285)):*
I see. Should we disallow mixing DDL and SQL? I'm thinking of

```SQL
BEGIN TRANSACTION;
CREATE INDEX foo ...;
SELECT * FROM tbl@foo;
COMMIT;
```

where the SELECT would fail since the index wouldn't have been created yet (?).

---

*[sql/schema_changer.go, line 396 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7liB7r0GyOZz6PGWwf:-K7liB7r0GyOZz6PGWwg:-2144546991) ([raw file](https://github.com/cockroachdb/cockroach/blob/7157e6fde24312ff5695637bcd22b605ad4215ef/sql/schema_changer.go#L396)):*
`return done && err == nil, err` to avoid returning non-zero values on error.

---

*[sql/schema_changer.go, line 470 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7liOayner-V31p8GiD:-K7liOaz63C2Rqr6kojq:1414813381) ([raw file](https://github.com/cockroachdb/cockroach/blob/7157e6fde24312ff5695637bcd22b605ad4215ef/sql/schema_changer.go#L470)):*
s/maa/mana/

---

*[sql/schema_changer.go, line 476 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7lioZiIdOq5bVezH-s:-K7lioZiIdOq5bVezH-t:-1596938471) ([raw file](https://github.com/cockroachdb/cockroach/blob/7157e6fde24312ff5695637bcd22b605ad4215ef/sql/schema_changer.go#L476)):*
if this ever gets called so late that `now` is very large compared to `execAfter`, then we're looking at a negative `d` and that sounds like it would panic in `NewTimer` below.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-:-K7lj9_zksIqfnRDnMq2:-622007672)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 18 of 19 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/executor.go, line 285 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7kU4mSVVzMxsRNWE-W:-K7m5rW6hCzMlDrkf9iC:-929027203) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc8c1f424af066a02845fddd7f72c05bfd449dd/sql/executor.go#L285)):*
I was originally thinking of going with the disallow model, but we later figured a way of not needing to do it. Of course we still need to educate our users that a schema change actually runs in multiple transactions, and the original command is only scheduling a schema change.

---

*[sql/schema_changer.go, line 396 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7liB7r0GyOZz6PGWwf:-K7m8SyKQJjwstwLZ4rQ:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/7157e6fde24312ff5695637bcd22b605ad4215ef/sql/schema_changer.go#L396)):*
done

---

*[sql/schema_changer.go, line 470 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7liOayner-V31p8GiD:-K7m7uFZ8P4Jkecuu-Lu:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/7157e6fde24312ff5695637bcd22b605ad4215ef/sql/schema_changer.go#L470)):*
done

---

*[sql/schema_changer.go, line 476 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7lioZiIdOq5bVezH-s:-K7m7xf5O7i25sIaGxjO:1487119487) ([raw file](https://github.com/cockroachdb/cockroach/blob/7157e6fde24312ff5695637bcd22b605ad4215ef/sql/schema_changer.go#L476)):*
I checked that it doesn't panic. In fact passing in negative values is legitimate

https://play.golang.org/p/tbRAQpYev7

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r10, 2 of 4 files at r11.
Review status: 18 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/backfill.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx18hDmCmZVKHwtNP:-K7mFWzKhMcmSCCjGoxp:773108201) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/backfill.go#L78)):*
That doesn't explain why there's a `break` here rather than a `continue`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 17 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/backfill.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx18hDmCmZVKHwtNP:-K7mHqrXis2veePB7J6e:-1942243589) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/backfill.go#L78)):*
Mutations are applied in a FIFO order. I've added that comment here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r13.
Review status: 18 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/backfill.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx18hDmCmZVKHwtNP:-K7mISxxZgjn4mbtZxyM:-1000117833) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/backfill.go#L78)):*
so it's also impossible that the first mutation doesn't belong to the thing being backfilled?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 18 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/backfill.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx18hDmCmZVKHwtNP:-K7mJ-0WP63YuC9DdO2l:2114363322) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/backfill.go#L78)):*
Not sure what you mean,  these schema changers are created through the gossip callback and there is no guarantee what the state of the actual table is before you read it in a transaction (which is what is happening here). Yes, the first mutation in the queue can be different from what gossip is telling you.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 18 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/backfill.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx18hDmCmZVKHwtNP:-K7mMIB5XhGKGkBhME0T:-188784521) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/backfill.go#L78)):*
What am I missing? For `break` to be safe here, the invariant must be that you're always looking for the first N mutations in the table; if you're looking for the a string of N mutations starting with the non-zeroth mutation, this will do the wrong thing.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 17 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/backfill.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411#-K7Wx18hDmCmZVKHwtNP:-K7mOnWa1acfHYJCO_lU:1196737656) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ec7a13ba3c1f8e14ef5a1d2c070bfaf11a8c0a/sql/backfill.go#L78)):*
This is meant to only look at the first set of mutations starting with index 0 that match the mutationID. I've added a comment about that, and copied that comment to the other sections of schema_changer.go that do the same.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r14.
Review status: 18 of 19 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3411)*
<!-- Sent from Reviewable.io -->
